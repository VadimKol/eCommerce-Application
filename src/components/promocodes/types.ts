import type { Cart } from '@commercetools/platform-sdk';

import type { Promocode } from '@/contexts/promocodes-context';

export interface PromocodesProps {
  addPromocode: (code: string) => Promise<void>;
  promocodes: Promocode[];
  cart: Cart;
}
