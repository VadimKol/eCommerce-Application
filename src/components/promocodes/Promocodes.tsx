import { type FormEvent, useRef } from 'react';
import { toast } from 'react-toastify';

import { Discounts } from '@/common/utils';
import { CustomButton } from '@/components/custom-button/Ð¡ustomButton';

import styles from './styles.module.scss';
import type { PromocodesProps } from './types';

export function Promocodes({ addPromocode, promocodes, cart }: PromocodesProps): JSX.Element {
  const promocodeInput = useRef<HTMLInputElement>(null);

  const appliedPromocodes = promocodes
    .filter(({ id }) => cart.discountCodes.map(({ discountCode }) => discountCode.id).includes(id))
    .map(({ name }) => name);

  const handlePromocode = (e: FormEvent): void => {
    e.preventDefault();
    if (promocodeInput.current?.value?.trim()) {
      const promocode = promocodeInput.current.value.trim();
      promocodeInput.current.value = '';
      if (!appliedPromocodes.includes(promocode)) {
        addPromocode(promocode)
          .then(() => toast('Successfully added Promocode', { type: 'success' }))
          .catch((error: Error) => toast(error.message, { type: 'error' }));
      } else {
        toast(`The discount code '${promocode}' already applied`, { type: 'error' });
      }
    }
  };

  return (
    <>
      {appliedPromocodes.map((promo) => (
        <p key={promo} className={styles.promocodes}>
          <span>{`"${promo}"`}</span>
          <span>{Discounts.find((item) => item.name === promo)?.desc}</span>
        </p>
      ))}
      <form className={styles.promobox} onSubmit={handlePromocode}>
        <p className={styles.promo_desc}>Enter promocode</p>
        <div className={styles.promoblock}>
          <input type="text" id="promocode" placeholder="Promocode" className={styles.promocode} ref={promocodeInput} />
          <CustomButton type="submit" aria-label="Promo-button">
            Apply
          </CustomButton>
        </div>
      </form>
    </>
  );
}
