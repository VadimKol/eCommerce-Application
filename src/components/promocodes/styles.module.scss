.promobox {
  margin-top: 16px;

  @media (width <= 1300px) {
    margin-top: 0.66em;
  }
}

.promo_desc {
  margin: 0 0 5px;
  font-size: 1.3rem;

  @media (width <= 1300px) {
    margin: 0;
    font-size: 0.91rem;
  }
}

.promoblock {
  display: flex;
  gap: 20px;
  justify-content: space-between;
  align-items: center;
  margin: 0 0 24px;

  @media (width <= 1300px) {
    margin: 0 0 16px;
  }
}

.promocode {
  box-sizing: border-box;
  width: 100%;
  height: 30px;
  padding: 0 5px;
  font-family: var(--primary-font);
  font-size: 1em;
  color: var(--primary-color);
  border: 1px solid var(--primary-color);
  border-radius: 6px;

  &:focus {
    border: 1px solid #000000;
  }

  &:focus-visible {
    outline: none;
  }
}

.promocodes {
  display: flex;
  justify-content: space-between;
  margin: 0;
  font-size: 1.2rem;
  color: var(--secondary-color);

  @media (width <= 1300px) {
    font-size: 0.84rem;
  }
}
