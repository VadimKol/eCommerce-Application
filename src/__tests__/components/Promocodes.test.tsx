import type { Cart } from '@commercetools/platform-sdk';
import { render } from '@testing-library/react';

import { Promocodes } from '@/components/promocodes/Promocodes';

it('renders correctly', () => {
  const cart: Cart = {
    id: '',
    version: 0,
    lineItems: [],
    customLineItems: [],
    totalPrice: {
      type: 'centPrecision',
      centAmount: 0,
      currencyCode: 'USD',
      fractionDigits: 0,
    },
    taxMode: 'Platform',
    taxRoundingMode: 'HalfEven',
    taxCalculationMode: 'LineItemLevel',
    inventoryMode: 'None',
    cartState: 'Active',
    shippingMode: 'Single',
    shipping: [],
    itemShippingAddresses: [],
    discountCodes: [],
    directDiscounts: [],
    refusedGifts: [],
    origin: 'Customer',
    createdAt: '',
    lastModifiedAt: '',
  };
  const { container } = render(<Promocodes promocodes={[]} addPromocode={jest.fn()} cart={cart} />);
  expect(container).toMatchSnapshot();
});
