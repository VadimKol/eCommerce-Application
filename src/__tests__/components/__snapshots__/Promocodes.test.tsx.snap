// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <form
    class="promobox"
  >
    <p
      class="promo_desc"
    >
      Enter promocode
    </p>
    <div
      class="promoblock"
    >
      <input
        class="promocode"
        id="promocode"
        placeholder="Promocode"
        type="text"
      />
      <button
        class="button button--primary"
        id=""
        type="submit"
      >
        Apply
      </button>
    </div>
  </form>
</div>
`;
